### ISSUE TITLE
Loss function scales

### ISSUE THREAD

**Issue by alessandro-montanari** [experiencor/basic-yolo-keras#127]
Hi,
I am training Full Yolo to detect two classes. I have about 13k images for training where 70% of them are of class 1 and 30% of class 0.
As you can see below I am getting very good results for class 1 while for class 0 the recall is not that great. The images for class 0 contain several small objects one next to each other while the images for class 1 contain only one object per image. 

<img width="442" alt="screen shot 2018-01-14 at 09 29 28" src="https://user-images.githubusercontent.com/3658219/34914179-7974598c-f90d-11e7-845f-b0cc1440f84e.png">

I was wondering if it is worth trying to play with the loss function scales (object_scale, coord_scale, ...) to increase the recall for class 0 but I don't know what would be a sensible way to do it.
Maybe I need to penalise more wrong boxes predictions and increase coord_scale?

Thanks

**Comment by experiencor** [comment]
@alessandro-montanari I see that your dataset is unbalanced. You can penalize different classess differently by adjusting ```self.class_wt``` in line 28 of ```frontend.py```. ```self.class_wt```is an one-dimentional array of class weights.

Other than that, you can try to use many small anchors of similar sizes in the same cell. It may help detect several small objects, which are near to one another. For example, the old list of anchor in increasing sizes is [anchor_1, anchor_2, anchor_3, anchor_4, anchor_5]. You can change to [anchor_1, anchor_1, anchor_1, anchor_2, anchor_3, anchor_4, anchor_5]. So now you have 7 anchors with 3 duplicate small anchors.

**Comment by alessandro-montanari** [comment]
@experiencor, I am trying to train with new anchor boxes generated with `gen_anchors.py`. I am trying one network with 5 anchors and one with 15. For now it seems that the val_loss is not much different than the traditional anchors but we will see once we test the model on the test set.

I will try to set the class weights as well. I will probably set the weight of class 0 to 2 and leave the one for class 1 at 1.

I was also thinking to apply learning rate decay on top of Adam optimiser. I have been reading around and some people say that it increased the accuracy, converging faster. An example is here with CapsNet https://github.com/XifengGuo/CapsNet-Keras/issues/9.
Do you have any thought on this?
Thanks a lot!

**Comment by experiencor** [comment]
@alessandro-montanari I have never try to apply learning rate decay on top of Adam optimiser. Curious to know your result!

**Comment by alessandro-montanari** [comment]
I already got a boost in performance with 5 anchors generated from the training set (see below) so I'm not sure I'll try the learning rate decay regime. If I do I'll let you know.

![screen shot 2018-01-17 at 16 28 33](https://user-images.githubusercontent.com/3658219/35053993-885afc1a-fba3-11e7-960e-26a4cc5abfb6.png)

**Comment by bicepjai** [comment]
how does custom loss function compare to https://github.com/allanzelener/YAD2K/blob/master/yad2k/models/keras_yolo.py#L152
